cpp/cmake_modules/SetupCxxFlags.cmake

```diff
@@ -274,29 +288,60 @@ if(BUILD_WARNING_FLAGS)
 endif(BUILD_WARNING_FLAGS)
 
 # Only enable additional instruction sets if they are supported
-if(CXX_SUPPORTS_AVX512 AND ARROW_SIMD_LEVEL STREQUAL "AVX512")
-  # skylake-avx512 consist of AVX512F,AVX512BW,AVX512VL,AVX512CD,AVX512DQ
-  set(CXX_COMMON_FLAGS "${CXX_COMMON_FLAGS} -march=skylake-avx512")
-elseif(CXX_SUPPORTS_AVX2 AND ARROW_SIMD_LEVEL STREQUAL "AVX2")
-  set(CXX_COMMON_FLAGS "${CXX_COMMON_FLAGS} -mavx2")
-elseif(CXX_SUPPORTS_SSE4_2 AND ARROW_SIMD_LEVEL STREQUAL "SSE4_2")
-  set(CXX_COMMON_FLAGS "${CXX_COMMON_FLAGS} -msse4.2")
+if(ARROW_USE_SIMD)
+  if(ARROW_SIMD_LEVEL STREQUAL "AVX512")
+    if(NOT CXX_SUPPORTS_AVX512)
+      message(FATAL_ERROR "AVX512 required but compiler doesn't support it.")
+    endif()
+    set(CXX_COMMON_FLAGS "${CXX_COMMON_FLAGS} ${ARROW_AVX512_FLAG}")
+    add_definitions(-DARROW_HAVE_AVX512 -DARROW_HAVE_AVX2)
+    add_definitions(-DARROW_HAVE_SSE4_2 -DARROW_HAVE_SSE2)
+  elseif(ARROW_SIMD_LEVEL STREQUAL "AVX2")
+    if(NOT CXX_SUPPORTS_AVX2)
+      message(FATAL_ERROR "AVX2 required but compiler doesn't support it.")
+    endif()
+    set(CXX_COMMON_FLAGS "${CXX_COMMON_FLAGS} ${ARROW_AVX2_FLAG}")
+    add_definitions(-DARROW_HAVE_AVX2)
+    add_definitions(-DARROW_HAVE_SSE4_2 -DARROW_HAVE_SSE2)
+  elseif(ARROW_SIMD_LEVEL STREQUAL "SSE4_2")
+    if(CXX_SUPPORTS_SSE4_2)
+      set(CXX_COMMON_FLAGS "${CXX_COMMON_FLAGS} ${ARROW_SSE4_2_FLAG}")
+      add_definitions(-DARROW_HAVE_SSE4_2 -DARROW_HAVE_SSE2)
+    elseif(CXX_SUPPORTS_SSE2)
```

Comment:
I'm not sure we should fall back here. Is it useful to support compilers with SSE2 but without SSE4.2 support?

Reply:
For Arrow supported compilers(gcc4.8+, clang-7+, msvc2015+), SSE2(actually SSE4.2) must be available.
There are some source code with "ifdef SSE2". I will remove SSE2 definition here and refine the code using it.